api/src/Controller/Department/ListDepartments.php

```diff
@@ -34,13 +34,38 @@
 
 namespace App\Controller\Department;
 
+use Slim\Container;
 use Slim\Http\Request;
 use Slim\Http\Response;
 
+use App\Service\DepartmentService;
+
 class ListDepartments extends BaseDepartment
 {
 
+    /**
+     * @var DepartmentService
+     */
+    protected $departmentService;
+
+
+    public function __construct(Container $container)
+    {
+        $this->departmentService = $container->get("DepartmentService");
+
+    }
+
+
+    public function __invoke(Request $request, Response $response, $params)
+    {
+        $output["type"] = "department_list";
+        $output["data"] = $this->departmentService->getAllDepartments();
+        return $response->withJson($output, 200);
+
+    }
+
 
+    // This function ignored with __construct and __invoke uncommented. Left here for now for reference.
```

Comment:
There is a LOT of backing work that is now being thrown away here that I have not been talked into yet.  If you look at the `BaseController` invoke you can see lots of error handling and automatic processing of the Json and includes all in one place.  You are abandoning all of that structure. 

Why are we not for a first step putting your new code into buildResource and returning the structure expected there.  That would make me much more comfortable with this change as it would be more incremental.

Reply:
>  If you look at the BaseController invoke you can see lots of error handling and automatic processing of the Json and includes all in one place. You are abandoning all of that structure.

> Why are we not for a first step putting your new code into buildResource and returning the structure expected there. That would make me much more comfortable with this change as it would be more incremental.

I'd like to answer both of these at the same time, because I think it will help illustrate the decision I made here to override invoke.

For error handling, I can see the reasoning in keeping this behavior. I think that this is something that would be best handled through the use of a middleware rather than an abstract class. It will help us move away from a model of inheritance in favor of one that leverages composition. In the case of this specific call, the only relevant HTTP Statuses are 200 (we got the data), 404 (we didn't have any data), and 500 (we goofed). I can certainly carry those over into this controller for now.

I'm not sure what you mean by automatic JSON processing exactly. I do see some logic that handles query parameters, which I think is valid. Is that what you're pointing out?

Can you help me understand why those were added initially, and whether we would need to keep them for this call? I know the obvious answer is so that we can paginate through data in case of _many results_, but for departments I didn't anticipate that there would be so many that it would be a problem.

As for skipping include logic, is there a concern about that here? I believe we are including all of the relevant information for Divisions and Departments with this request, including enough about parent and fallback departments that API consumers can access what they need from the rest of the response. I'd be willing to concede that we are not returning the full object for parent and fallback department, but again it felt like it was not required because that data exists elsewhere in the response.

Regarding putting new code into `buildResource`, I think that largely most of the underlying code wasn't required to fetch the data for this endpoint. I will concede that battle-tested error handling and query parameter processing would be nice to have. However, I would need to eliminate the includes altogether because I would not want those to be run, ever, for this endpoint. It might be that I could replace includes in the `ListDepartments` controller with an empty array.

For better or worse, it felt cleaner to have a fresh path for this code. Most of the underlying functionality has been replaced, so I didn't want to add confusion by leveraging the abstract function because a lot of the request processing didn't feel applicable. It felt best to start pulling the bandage off, if that makes sense.

I agree that it might be more incremental to use `buildResource`, but a number of parameters would've been flat out ignored anyway unless I implemented them in the service and repository layer, which could be done and isn't impossible, but based on the performance it seemed fine if we didn't.

Reply:
Just as an overview, I really like this design and path forward but I am being very incremental and cautious at our implementation.  There actually has been a LOT of design and thought put into this and while the structure is a bit of a mess I want to try to make changes more incrementally and with the same or more thought and planning.

> > If you look at the BaseController invoke you can see lots of error handling and automatic processing of the Json and includes all in one place. You are abandoning all of that structure.
> 
> > Why are we not for a first step putting your new code into buildResource and returning the structure expected there. That would make me much more comfortable with this change as it would be more incremental.
> 
> I'd like to answer both of these at the same time, because I think it will help illustrate the decision I made here to override invoke.
> 
> For error handling, I can see the reasoning in keeping this behavior. I think that this is something that would be best handled through the use of a middleware rather than an abstract class. It will help us move away from a model of inheritance in favor of one that leverages composition. In the case of this specific call, the only relevant HTTP Statuses are 200 (we got the data), 404 (we didn't have any data), and 500 (we goofed). I can certainly carry those over into this controller for now.
> 

Could you help me see the middleware approach here?  Maybe we implement this middleware first because then that logic can be removed from the existing code path. 

> I'm not sure what you mean by automatic JSON processing exactly. I do see some logic that handles query parameters, which I think is valid. Is that what you're pointing out?
> 
> Can you help me understand why those were added initially, and whether we would need to keep them for this call? I know the obvious answer is so that we can paginate through data in case of _many results_, but for departments I didn't anticipate that there would be so many that it would be a problem.
> 
> As for skipping include logic, is there a concern about that here? I believe we are including all of the relevant information for Divisions and Departments with this request, including enough about parent and fallback departments that API consumers can access what they need from the rest of the response. I'd be willing to concede that we are not returning the full object for parent and fallback department, but again it felt like it was not required because that data exists elsewhere in the response.
> 

I am looking at the `__invoke` implementation in 'BaseController'. The base goal was to have very consistent reply construction such as ensuring that ever returned resource has an `id` and a `type` and that lists are consistently paged.  Also the goal was to remove as much of all that from the individual endpoint implementations as possible. 

I need to dig more into the include logic because you are insistent that the return is the same and I am not seeing that in the code so maybe I am missing something.  I would really like this change be unchanging to the schema. If we want to change the schema for departments or staff that is fine but that is not part of a backend change like this. It should be done independently. 

> Regarding putting new code into `buildResource`, I think that largely most of the underlying code wasn't required to fetch the data for this endpoint. I will concede that battle-tested error handling and query parameter processing would be nice to have. However, I would need to eliminate the includes altogether because I would not want those to be run, ever, for this endpoint. It might be that I could replace includes in the `ListDepartments` controller with an empty array.
> 
> For better or worse, it felt cleaner to have a fresh path for this code. Most of the underlying functionality has been replaced, so I didn't want to add confusion by leveraging the abstract function because a lot of the request processing didn't feel applicable. It felt best to start pulling the bandage off, if that makes sense.
> 
> I agree that it might be more incremental to use `buildResource`, but a number of parameters would've been flat out ignored anyway unless I implemented them in the service and repository layer, which could be done and isn't impossible, but based on the performance it seemed fine if we didn't.

I am torn here because I understand you sentiment, yet when making changes like this to a core part of the system I am far less of a "tearing bandage" person and more of a "progressive incremental change" person.  Like I said above I would really want this change to be schema identical. 

Maybe this is just because I am the author here but even if not a scrap of the old code exists after this restructure I would be far more comfortable taking this in smaller steps replacing it bit by bit than a whole hog reimplementation from the ground up. 

I would be willing to be more relaxed about this for something like the `Staff` submodule, but when we are doing `Cycles` and `Events` and `Departments` we are touch things that touch many other things and we would have to be sure everything still works together at each step, and not just because it is still using the old code in parallel.

Reply:
> I want to try to make changes more incrementally and with the same or more thought and planning.

I think that's fair. The way I proposed these changes by completely bypassing the stuff from `BaseController` is a bit drastic, compared to a more incremental approach. I think it can probably be achieved incrementally by responding with the data I've got from the new pattern in `buildResource`, and make sure I prevent any `includes` from running unless we really want to run them. 

My preference for this call is to avoid them, for two reasons:
- We would lose out on the performance benefits of this call if we kept them, even with re-organizing the code a bit.
- I think the only code currently using this endpoint is on the ConCom List page. This isn't removing the call in the `BaseController` so other areas won't be impacted by this until we move them to the new pattern
- The data is about as schema-identical as we can reasonably get

I have attached sample responses for the original and the updated request.
[original_sample.json](https://github.com/CON-In-A-Box/CIAB-Portal/files/13506729/original_sample.json)
[updated_sample.json](https://github.com/CON-In-A-Box/CIAB-Portal/files/13506730/updated_sample.json)

In both we have `id`, `name`, `child_count`, `email`. 

In the old, `parent` also includes _the entire parent object_ which is present earlier in the response output. If that parent had `fallback` values, you'd see the entire chain of `fallback` departments in full.

In the new, `parent` simply contains `id` and `name`. `fallback` in the new is similarly just `id` and `name`.

In this specific API request, we will always have a full list of departments. We always populate parent and fallback ID and Name if present. It felt redundant to walk any further down the list, if that makes sense. If we look at a department and it has a parent, we have the info we need to find the parent department if we want more detail. In that sense, we aren't missing any data.

I deliberately chose to ignore the case where we look up a single department by ID. In a call like that, we could certainly provide a bit more detail about the parent or fallback, but in a request that gets us a list it felt less applicable.

I think for the other areas you mention that I agree, it will be good to make sure they work at every step. It seems that one of the primary concerns (unless I've misunderstood) is to make sure the UI has everything it needs from an API call, but I also want to make sure we don't conflate that with "API must return all DB fields exactly", either, because it may not be required.

Reply:
> > I want to try to make changes more incrementally and with the same or more thought and planning.
> 
> I think that's fair. The way I proposed these changes by completely bypassing the stuff from `BaseController` is a bit drastic, compared to a more incremental approach. I think it can probably be achieved incrementally by responding with the data I've got from the new pattern in `buildResource`, and make sure I prevent any `includes` from running unless we really want to run them.
> 
> My preference for this call is to avoid them, for two reasons:
> 
> * We would lose out on the performance benefits of this call if we kept them, even with re-organizing the code a bit.
> * I think the only code currently using this endpoint is on the ConCom List page. This isn't removing the call in the `BaseController` so other areas won't be impacted by this until we move them to the new pattern
> * The data is about as schema-identical as we can reasonably get
> 

We found other places using the `Departments` entry point so we will have to make sure that everything works. I am willing to rethink how includes are done but I would like it to be API wide policy and not bit and pieces.  I may have some time to do a full API audit but we can see. 

For the `Staff` specific stuff we can make changes where it is not used anywhere else for now. So I am not feeling as twitchy about this.  

> I have attached sample responses for the original and the updated request. [original_sample.json](https://github.com/CON-In-A-Box/CIAB-Portal/files/13506729/original_sample.json) [updated_sample.json](https://github.com/CON-In-A-Box/CIAB-Portal/files/13506730/updated_sample.json)
> 
> In both we have `id`, `name`, `child_count`, `email`.
> 
> In the old, `parent` also includes _the entire parent object_ which is present earlier in the response output. If that parent had `fallback` values, you'd see the entire chain of `fallback` departments in full.
> 
> In the new, `parent` simply contains `id` and `name`. `fallback` in the new is similarly just `id` and `name`.
> 
> In this specific API request, we will always have a full list of departments. We always populate parent and fallback ID and Name if present. It felt redundant to walk any further down the list, if that makes sense. If we look at a department and it has a parent, we have the info we need to find the parent department if we want more detail. In that sense, we aren't missing any data.

I can  an argument to limit includes to maybe 2 layers at most unless maybe specifically asked for.  I would be surprised if we are relying on anything more than 2 layers deep but we would have to audit. 

> 
> I deliberately chose to ignore the case where we look up a single department by ID. In a call like that, we could certainly provide a bit more detail about the parent or fallback, but in a request that gets us a list it felt less applicable.
> 
> I think for the other areas you mention that I agree, it will be good to make sure they work at every step. It seems that one of the primary concerns (unless I've misunderstood) is to make sure the UI has everything it needs from an API call, but I also want to make sure we don't conflate that with "API must return all DB fields exactly", either, because it may not be required.

Reply:
> > We found other places using the Departments entry point so we will have to make sure that everything works. I am willing to rethink how includes are done but I would like it to be API wide policy and not bit and pieces. I may have some time to do a full API audit but we can see.
> 
> I did find that the other parts of the code using the List Departments API are only using department ID and name, which are both provided in the response.
> 
> As far as this part:
> 
> > I am willing to rethink how includes are done but I would like it to be API wide policy and not bit and pieces.
> 
> I want to make sure I understand and am on the same page. When you talk about includes in this case, are you talking about specifically the logic in the `BaseController` that we are looking to re-visit? Or are you talking about the extra data attached to an entity such as having a Department that also has a Parent and Fallback Department included?
> 

Logic in the `BaseController` or wherever it ends up moving to as we gut that module.

> And then by API wide policy, are you meaning "something generic that can handle any given data"? If that's the case, I think it would be much more difficult than just a single PR to accomplish. If it's not the case, can you clarify what you mean?
> 

Yes, and agreed.  But I want to be moving clearly in that direction and not in a code-duplications every resource does it themselves sort of way.

> > I can an argument to limit includes to maybe 2 layers at most unless maybe specifically asked for. I would be surprised if we are relying on anything more than 2 layers deep but we would have to audit.
> 
> IMO, for this particular case, it probably isn't necessary to even go 2 layers deep unless you mean "Department + its Immediate Parent (no parent or fallback chaining) and Immediate Fallback (no parent or fallback chaining)"
> 
> It's probably also not worth considering allowing a query parameter to specifically ask for more. I'm guessing there are other ways around this if we're needing to go any farther than the baseline and it might make more sense to address it when we encounter that scenario rather than trying to plan for it up front. I think we will have the flexibility to make this kind of change down the road without a lot of invested effort.

Thinking here.  I think I can let this pass and just boil a bit as we work on other areas to make vision into reality.

Reply:
>  But I want to be moving clearly in that direction and not in a code-duplications every resource does it themselves sort of way.

Yeah, I think we want the same thing then. However I also _strongly_ encourage you to consider a scenario where we evaluate a proper pattern _after_ we've had a chance to move a few of these.

Right now, I'd much rather address the immediate performance issues, and _then_ focus on "OK, what's the best way to establish a proper pattern".

Because we will need to handle nested objects, I think solving this in a generic way is going to be something that we have to solve by iterating.